bin/submitty_autograding_shipper.py

```diff
@@ -0,0 +1,417 @@
+        except Exception as e:
+            grade_items_logging.log_message(message="ERROR: Could not retrieve the file from the foreign machine "+str(e))
+            print("ERROR: Could not retrieve the file from the foreign machine.\nERROR: {0}".format(e))
+            success = False
+        finally:
+            os.close(fd1)
+            os.close(fd2)
+            sftp.close()
+            ssh.close()
+            if not success:
+                return False
+    # archive the results of grading
+    try:
+        grade_item.unpack_grading_results_zip(which_machine,which_untrusted,local_results_zip)
+    except:
+        grade_items_logging.log_message(jobname=next_to_grade,message="ERROR: Exception when unpacking zip")
+        with contextlib.suppress(FileNotFoundError):
+            os.remove(local_results_zip)
+
+    with contextlib.suppress(FileNotFoundError):
+        os.remove(local_done_queue_file)
+    return True
+
+
+# ==================================================================================
+def grade_queue_file(which_machine,which_untrusted,queue_file):
+    """
+    Oversees the autograding of single item from the queue
+
+    :param queue_file: details of what to grade
+    :param which_machine: name of machine to send this job to (might be "localhost")
+    :param which_untrusted: specific untrusted user for this autograding job
+    """
+
+    my_dir,my_file=os.path.split(queue_file)
+    pid = os.getpid()
+    directory = os.path.dirname(os.path.realpath(queue_file))
+    name = os.path.basename(os.path.realpath(queue_file))
+    grading_file = os.path.join(directory, "GRADING_" + name)
+
+    #TODO: breach which_machine into id, address, and passphrase.
+    which_machine="localhost"
+
+    try:
+        # prepare the job
+        shipper_counter=0
+        while not prepare_job(which_machine, which_untrusted, my_dir, queue_file):
+            shipper_counter = 0
+            time.sleep(1)
+            if shipper_counter >= 10:
+                prints(which_machine, which_untrusted, "shipper prep loop: ",queue_file)
+                shipper_counter=0
+
+        # then wait for grading to be completed
+        shipper_counter=0
+        while not unpack_job(which_machine, which_untrusted, my_dir, queue_file):
+            shipper_counter+=1
+            time.sleep(1)
+            if shipper_counter >= 10:
+                print (which_machine,which_untrusted,"shipper wait for grade: ",queue_file)
+                shipper_counter=0
+
+    except Exception as e:
+        print ("ERROR attempting to grade item: ", queue_file, " exception=",str(e))
+        grade_items_logging.log_message(message="ERROR attempting to grade item: " + queue_file + " exception " + repr(e))
+
+    # note: not necessary to acquire lock for these statements, but
+    # make sure you remove the queue file, then the grading file
+    try:
+        os.remove(queue_file)
+    except Exception as e:
```

Comment:
what exceptions are we worried about throwing here? OSError (trying to delete directory), PermissionError, or FileNotFoundError?